### PR TITLE
fix(glob): fix patterns starting with * #8817

### LINKED ISSUE #8817: Glob patterns starting with "*." and having multiple wildcards fails
### What version of Bun is running?

1.0.26+c75e768a6

### What platform is your computer?

Linux 6.5.0-15-generic x86_64 x86_64

### What steps can reproduce the bug?

Considering following project structure:
```
project/
├ test/
│ └ example.test.ts
├ index.ts
├ package.json
└ tsconfig.json
```
With following `index.ts` content:
```ts
import { Glob } from "bun"

for await (const f of new Glob('*.test.*').scan({ cwd: "test" })) {
  console.log(f)
}
console.log("done")
```

### What is the expected behavior?

According to [Glob API](https://bun.sh/docs/api/glob), I would expect the following output:
```bash
$ bun run index.ts
example.test.ts
done
```

### What do you see instead?

No file matches the given pattern
```bash
$ bun run index.ts
done
```

### Additional information

As a workaround, escaping the first dot seems to work correctly
```js
new Glob('*\\.test.*')
```

### PR DESCRIPTION
**Pull request by zackradisic**
### What does this PR do?

Fixes #8817, the issue was properly applying the wildcard-filepath optimization. I added tests to make sure its used properly

### CODE REVIEW

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Shell leak test is failing

**Comment by Jarred-Sumner** [comment]
there are merge conflicts

**Comment by Jarred-Sumner** [comment]
74 failing scan tests

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Test failures

File: src/glob.zig [hunk 1]
```diff
@@ -673,6 +693,10 @@ pub fn GlobWalker_(
             this.error_on_broken_symlinks = error_on_broken_symlinks;
             this.only_files = only_files;
 
+            if (bun.Environment.allow_assert) {
+                this.debugPatternComopnents();
```
Comment (jdalton) [inline src/glob.zig:697]: ☝️ Does this make a pretty debug log?

File: src/glob.zig [hunk 2]
```diff
@@ -1108,7 +1132,20 @@ pub fn GlobWalker_(
...
+                                //
+                                // This is not a complete check, there can be
+                                // false negatives, but that's okay, it just
+                                // means we don't apply the optimization.
+                                //
+                                // We also don't need to look for the `!` token,
+                                // because that only applies negation if at the
+                                // beginning of the string.
+                                '[', '{', '?', '*' => break :out_of_check_wildcard_filepath,
                                 else => {},
```
Comment (jdalton) [inline src/glob.zig:1149]: Love the code comments here ❤️

File: src/glob.zig [hunk 3]
```diff
@@ -200,7 +200,8 @@ pub fn GlobWalker_(
             fds_open: if (bun.Environment.allow_assert) usize else u0 = 0,
 
             pub fn init(this: *Iterator) !Maybe(void) {
-                const path_buf: *[bun.MAX_PATH_BYTES]u8 = &this.walker.pathBuf;
+                log("Iterator init pattern={s}", .{this.walker.pattern});
+                var path_buf: *[bun.MAX_PATH_BYTES]u8 = &this.walker.pathBuf;
```
Comment (jdalton) [inline src/glob.zig:204]:
```suggestion
                var path_buf: *bun.PathBuffer = &this.walker.pathBuf;
```

**Review by jdalton — APPROVED** [review]
Pretty scoped debug, code comments, unit tests. What's not to like. Minor nit for extra expressiveness but I dig!

**Comment by Jarred-Sumner** [comment]
Merging this so we can get it in the release

**Comment by leaumar** [comment]
Thanks!

**Pull request merged**